### ISSUE TITLE
Investigate and fix definition search and loading performance issues in codebase UI

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#2045]
This meta-ticket is tracking work to improve the performance of fuzzy find and definition loading. The current performance issues affect the codebase UI used by Unison Share and the local codebase UI.

1. Fuzzy find linearly scans all names in the codebase and performance is poor on large codebases (it takes seconds). There's probably some constant factor improvements to be made, but we should introduce an index of some sort and then cache that in the codebase server backend for the latest branch. A suffix trie index can be used to answer `isInfixOf` checks in time proportional to just the query length (not the corpus size) or there might be an even fancier indexing structure we could use that could answer isSubsequenceOf queries efficiently.
2. Definition loading is slow on large codebases as well (5-10s); we suspect but haven't confirmed it's due to building the `Names` and `PrettyPrintEnv` which have to do linear work in the size of the namespace. These can be cached in the AppState and there's WIP on this in https://github.com/unisonweb/unison/pull/2023
3. Unison Share backend performance degrades with concurrent requests. It seems like there's either a) a simple misconfiguration of the backend, like it's not compiled with `-threaded` and using multiple CPUs and/or b) there's some unintentional contention in the implementation.

__Todo:__

- [ ] Double check that it's actually being run multithreaded with -O2
  - [ ] Potentially GitHub action needs tweaking https://docs.haskellstack.org/en/stable/yaml_configuration/ to pass optimized flag
  - [ ] Make sure passing `-threaded` and setting number of CPUs correctly - both need to be set

**Comment by hojberg** [comment]
This seems related to https://github.com/unisonweb/unison/issues/2109 - the 2 issues should perhaps be merged? @pchiusano @stew

**Comment by puffnfresh** [comment]
A bunch of requests get created as I type:

```
b
ba
bas
base
base.
base.L
base.Li
base.Lis
base.List
base.List.
base.List.h
base.List.he
base.List.hea
base.List.head
```

This overloads the backend. Even though a single request of "base.List.head" can take about 2 seconds, my searches take over 20 seconds due to being overloaded.

I think the UI should cancel the requests and the backend should cancel the threads.

**Comment by hojberg** [comment]
@puffnfresh good point.

We've been debouncing the Finder requests from the UI on Unison Share, but had kept each request for Unison Local since at the time it returned quick enough. Thats not the case anymore. It seems like it's time to enable that.

**Comment by pchiusano** [comment]
Did something happen recently that tanked performance? I feel like after https://github.com/unisonweb/unison/pull/2292 (see the videos there), search felt basically instantaneous (see the video in that PR), but I just tried trunk and it's back to being super slow.

**Comment by hojberg** [comment]
@pchiusano the introduction of using the `rootBranch` query param seems to have hurt things a lot.

**Comment by pchiusano** [comment]
@hojberg I spotted something in `Backend.hs`:

```Haskell
resolveBranchHash
  :: Monad m => Maybe Branch.Hash -> Codebase m v Ann -> Backend m (Branch m)
resolveBranchHash h codebase = case h of
  Nothing    -> getRootBranch codebase
  Just bhash -> do
    mayBranch <- lift $ Codebase.getBranchForHash codebase bhash
    mayBranch ?? NoBranchForHash bhash
```

This is calling `Codebase.getBranchFromHash` which currently loads the branch from disk every time. This will kill performance.

What I'd do is rename `Codebase.getBranchForHash` to `getBranchForHashImpl`, and then have `getBranchForHash` check if the requested hash is one of the descendants of the root branch (maybe limit depth of search), which is cached in memory. If it is, just return it directly. Otherwise it can call the `getBranchForHashImpl`.

I tested on the topic/substring-fzf branch, and it's slow on a big enough namespace where there's like multiple forks of base, so even with the above I think more work is needed.

Another thing that should probably be done that I thought of is deduping search results by hash before generating the results. This should help performance and also it's a better DX anyway - I don't need to see multiple rows for the same definition in the finder.
